frame/oracle/src/lib.rs

```diff
@@ -940,51 +947,86 @@ pub mod pallet {
 			});
 		}
 
-		// REVIEW: indexing
-		#[allow(clippy::indexing_slicing)] // to get CI to pass
 		pub fn get_twap(
 			asset_id: T::AssetId,
-			mut price_weights: Vec<T::PriceValue>,
+			twap_window: usize,
 		) -> Result<T::PriceValue, DispatchError> {
-			let precision: T::PriceValue = 100_u128.into();
 			let historical_prices = Self::price_history(asset_id);
 
-			// add an extra to account for current price not stored in history
-			ensure!(historical_prices.len() + 1 >= price_weights.len(), Error::<T>::DepthTooLarge);
-
-			let sum = Self::price_values_sum(&price_weights);
-			ensure!(sum == precision, Error::<T>::MustSumTo100);
-
-			let last_weight = price_weights.pop().unwrap_or_else(|| 0_u128.into());
-			ensure!(last_weight != 0_u128.into(), Error::<T>::ArithmeticError);
+			ensure!(twap_window <= historical_prices.len() + 1, Error::<T>::DepthTooLarge);
 
-			let mut weighted_prices = price_weights
+			let mut prices: Vec<_> = historical_prices
 				.iter()
+				.rev()
 				.enumerate()
-				.map(|(i, weight)| {
-					weight
-						.mul(
-							historical_prices[historical_prices.len() - price_weights.len() + i]
-								.price,
-						)
-						.div(precision)
+				.filter_map(|(i, price)| if i < twap_window { Some(price) } else { None })
+				.rev()
+				.collect();
+			let current_price =
+				Prices::<T>::try_get(asset_id).map_err(|_| Error::<T>::PriceNotFound)?;
+			prices.push(&current_price);
+
+			let weights: Vec<u128> = prices
+				.windows(2)
+				.map(|w| {
+					w.get(0)
+						.and_then(|a| w.get(1).map(|b| (a, b)))
+						.and_then(|(a, b)| b.block.checked_sub(&a.block))
+						.unwrap_or_default()
+						.unique_saturated_into()
 				})
-				.collect::<Vec<_>>();
-			let current_price = Self::prices(asset_id);
-			let current_weighted_price = last_weight.mul(current_price.price).div(precision);
+				.collect();
 
-			weighted_prices.push(current_weighted_price);
+			let weights_sum: u128 = weights.iter().sum();
```

Comment:
We can add an `ensure` to assert that `weight_sum = latest_block - earliest block in the window`